syne_tune/optimizer/schedulers/transfer_learning/quantile_based/quantile_based_searcher.py

```diff
@@ -0,0 +1,157 @@
+    y = np.concatenate(ys, axis=0)
+    return X, y
+
+
+def fit_model(
+        config_space,
+        transfer_learning_evaluations,
+        normalization: str,
+        max_fit_samples: int,
+        random_state,
+        model=xgboost.XGBRegressor()
+):
+    model_pipeline = BlackboxSurrogate.make_model_pipeline(
+        configuration_space=config_space,
+        fidelity_space={},
+        model=model,
+    )
+    X, y = extract_input_output(transfer_learning_evaluations, normalization, random_state=random_state)
+    with catchtime("time to fit the model"):
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=random_state)
+        X_train, y_train = subsample(X_train, y_train, max_samples=max_fit_samples, random_state=random_state)
+        model_pipeline.fit(X_train, y_train)
+
+        # compute residuals (num_metrics,)
+        sigma_train = eval_model(model_pipeline, X_train, y_train)
+
+        # compute residuals (num_metrics,)
+        sigma_val = eval_model(model_pipeline, X_test, y_test)
+
+    return model_pipeline, sigma_train, sigma_val
+
+
+def eval_model(model_pipeline, X, y):
+    # compute residuals (num_metrics,)
+    mu_pred = model_pipeline.predict(X)
+    if mu_pred.ndim == 1:
+        mu_pred = mu_pred.reshape(-1, 1)
+    res = np.std(y - mu_pred, axis=0)
+    return res.mean()
+
+
+def subsample(X_train, z_train, max_samples: int = 10000, random_state: np.random.RandomState = None):
+    assert len(X_train) == len(z_train)
+    X_train.reset_index(inplace=True)
+    if max_samples is not None and max_samples < len(X_train):
+        if random_state is None:
+            random_indices = np.random.permutation(len(X_train))[:max_samples]
+        else:
+            random_indices = random_state.permutation(len(X_train))[:max_samples]
+        X_train = X_train.loc[random_indices]
+        z_train = z_train[random_indices]
+    return X_train, z_train
+
+
+class QuantileBasedSurrogateSearcher(SearcherWithRandomSeed):
+
+    def __init__(
+            self,
+            config_space: Dict,
+            metric: str,
+            transfer_learning_evaluations: Dict[str, TransferLearningTaskEvaluations],
+            mode: Optional[str] = None,
+            max_fit_samples: int = 100000,
+            normalization: str = "gaussian",
+            random_seed: Optional[int] = None,
+    ):
+        """
+        Implement the transfer-learning method:
+        A Quantile-based Approach for Hyperparameter Transfer Learning.
+        David Salinas, Huibin Shen, Valerio Perrone. ICML 2020.
+        This is the Copula Thompson Sampling approach described in the paper where a surrogate is fitted on the
+        transfer learning data to predict mean/variance of configuration performance given a hyperparameter.
+        The surrogate is then sampled from and the best configurations are returned as next candidate to evaluate.
```

Comment:
sampled from ?

Reply:
Yes, the surrogate provide mean/sigma estimate for each configurations.